### DISCUSSION TITLE
Client policies: Migration strategy for existing clients

### DISCUSSION THREAD

**Discussion by danielFesenmeyer** [keycloak/keycloak#12405] [Q&A]
Hi,

We consider to make use of the "Client policies" feature in order to make sure that our customers' clients are secured appropriately.
Actually we have quite many existing clients, and lots of those don't conform to the PKCE enforcer (we have not checked other enforcers yet).
Ideally, we would like to make sure that new clients are created the proper way (i.e. conform to the policy), whereas existing clients still work as before, until the corresponding applications have been migrated (e.g. to support PKCE).

But the question is how we can achieve this?

The first thing I thought of was to configure a _client-updater-context_ condition for the policy (with all available _Update Client Contexts_). This way, the policy would only be applied, in case of client creation or update, not on any other events such as authorization requests. The PKCE enforcer would be configured with auto-configure disabled, to avoid clients to be updated with the PKCE setting, in case something unrelated is changed (like the displayName) - cause this could break existing applications.
But the downside is that with this configuration, via the Admin UI (both V1 and V2), clients can no longer be created - because the UI always first creates an "incomplete" client, which can then be further configured by the user. With auto-configure=false, the creation of this "incomplete" client fails. Is this a bug or expected behavior? Wouldn't it be better if the whole client would be created with a single request?

To fix the UI's client creation problem, we could set auto-configure=true for the PKCE enforcer. But as mentioned above, this would lead to automatic configuration of the PKCE setting, in case of any unrelated update of the client, and might break existing applications.

Another idea would be to implement our own condition, which just applies to client creation (NOT update). This condition could be replaced with an _any-client_ condition, as soon as all applications have been migrated.

WDYT?

**Comment by tnorimat** [comment]
Hello @danielFesenmeyer ,
 
Your idea will work well. I have also the following 2 ideas.

- an existing app: does not support PKCE, so PKCE is not enforced
- an app to be created: support PKCE, so PKCE is enforced


1. Distinguishing an existing app and an app to be created

If an existing app have some characteristics that an app to be created does not have, we can use it.
For example, if an existing app is assigned a role, `ClientRolesCondition` can be used.

- creating a policy that consists of `ClientUpdaterContextCondition` and `ClientRolesCondition`.
- setting Roles of `ClientRolesCondition` to the role assigned to only an existing app.
- setting Negative Logic of `ClientRolesCondition` to ON, which inverts the results of the condition.
- creating a profile that consists of `PKCEEnforcerExecutor`.
- setting Auto Configure of `PKCEEnforcerExecutor`to ON.
- making the policy refer to the profile.

On Admin UI, if we modify an existing app, the policy is not applied.
On Admin UI, if we create an app, the policy is applied, and the app is successfully created with valid PKCE setting enforced.
On Admin UI, if we modify an newly created app, the policy is applied, and the app is successfully modified.
On Admin UI, if we migrate an existing app to the one that support PKCE, we first drop the role from the app. Next, we modify the app. Then the app is successfully modified with valid PKCE setting enforced.

If no such role is assigned to an existing app, we would find out other characteristics that only an existing app has, and write our own condition to use the characteristics.

2. Distinguishing an administrator for managing existing apps and for apps to be created

We setup two administrator, the one is for managing existing apps and the other is for managing apps o be created.
These administrator can be distinguished by some characteristics, like a group and a role.

For example, the former belongs "legacy" group while the latter not.

- creating a policy that consists of `ClientUpdaterSourceGroupsCondition`.
- setting Groups of `ClientUpdaterSourceGroupsCondition` to "legacy".
- setting Negative Logic of `ClientUpdaterSourceGroupsCondition` to ON, which inverts the results of the condition.
- creating a profile that consists of `PKCEEnforcerExecutor`.
- setting Auto Configure of `PKCEEnforcerExecutor`to ON.
- making the policy refer to the profile.

On Admin UI with the former admin, if we modify an existing app, the policy is not applied.
On Admin UI with the latter admin, if we create an app, the policy is applied, and the app is successfully created with valid PKCE setting enforced.
On Admin UI with the latter admin, if we modify an newly created app, the policy is applied, and the app is successfully modified.
On Admin UI with the latter admin, if we migrate an existing app to the one that support PKCE, we modify the app. Then the app is successfully modified with valid PKCE setting enforced.
